crates/optimism/chainspec/src/lib.rs

```diff
@@ -34,21 +34,58 @@ use reth_chainspec::{
 };
 use reth_ethereum_forks::{ChainHardforks, EthereumHardfork, ForkCondition};
 use reth_network_peers::NodeRecord;
+use reth_optimism_forks::OptimismHardfork;
 use reth_primitives_traits::Header;
 
+/// These are the masks for the base fee denominator and elasticity in the nonce, post Holocene
+const DENOMINATOR_MASK: B64 = b64!("FFFFFFFF00000000");
+const ELASTICITY_MASK: B64 = b64!("00000000FFFFFFFF");
+
 /// OP stack chain spec type.
 #[derive(Debug, Clone, Deref, Into, Constructor, PartialEq, Eq)]
 pub struct OpChainSpec {
     /// [`ChainSpec`].
     pub inner: ChainSpec,
 }
 
+/// Fee trait for OP chain specs.
+pub trait Fee {
+    /// Read from parent to determine the base fee for the next block
+    fn next_block_base_fee(&self, parent: &Header, timestamp: u64) -> U256;
+}
+
 /// Returns the signature for the optimism deposit transactions, which don't include a
 /// signature.
 pub fn optimism_deposit_tx_signature() -> Signature {
     Signature::new(U256::ZERO, U256::ZERO, Parity::Parity(false))
 }
 
+impl Fee for OpChainSpec {
+    fn next_block_base_fee(&self, parent: &Header, timestamp: u64) -> U256 {
+        let is_holocene =
+            self.inner.is_fork_active_at_timestamp(OptimismHardfork::Holocene, timestamp);
+        let not_first =
+            self.inner.is_fork_active_at_timestamp(OptimismHardfork::Holocene, parent.timestamp);
+        // If we are in the Holocene, we need to use the base fee params from the parent block's
+        // nonce Else, use the base fee params from chainspec
+        if is_holocene && not_first {
```

Comment:
I think we should probably change this to `is_holocene && parent.nonce != B64::ZERO`.

Reply:
See [this section](https://specs.optimism.io/protocol/holocene/exec-engine.html#execution) of the spec.